controller/api/destination/watcher/ip_watcher.go

```diff
@@ -0,0 +1,372 @@
+)
+
+// NewIPWatcher creates an IPWatcher and begins watching the k8sAPI for service
+// changes.
+func NewIPWatcher(k8sAPI *k8s.API, endpoints *EndpointsWatcher, log *logging.Entry) *IPWatcher {
+	iw := &IPWatcher{
+		publishers: make(map[string]*serviceSubscriptions),
+		endpoints:  endpoints,
+		k8sAPI:     k8sAPI,
+		log: log.WithFields(logging.Fields{
+			"component": "ip-watcher",
+		}),
+	}
+
+	k8sAPI.Svc().Informer().AddIndexers(cache.Indexers{podIPIndex: func(obj interface{}) ([]string, error) {
+		if svc, ok := obj.(*corev1.Service); ok {
+			return []string{svc.Spec.ClusterIP}, nil
+		}
+		return []string{""}, fmt.Errorf("object is not a service")
+	}})
+
+	k8sAPI.Svc().Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc:    iw.addService,
+		DeleteFunc: iw.deleteService,
+		UpdateFunc: func(before interface{}, after interface{}) {
+			iw.deleteService(before)
+			iw.addService(after)
+		},
+	})
+
+	k8sAPI.Pod().Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc:    iw.addPod,
+		DeleteFunc: iw.deletePod,
+		UpdateFunc: func(_ interface{}, obj interface{}) {
+			iw.addPod(obj)
+		},
+	})
+
+	return iw
+}
+
+////////////////////////
+/// IPWatcher ///
+////////////////////////
+
+// Subscribe to an authority.
+// The provided listener will be updated each time the address set for the
+// given authority is changed.
+func (iw *IPWatcher) Subscribe(clusterIP string, port Port, listener EndpointUpdateListener) error {
+	iw.log.Infof("Establishing watch on service cluster ip [%s:%d]", clusterIP, port)
+	ss := iw.getOrNewServiceSubscriptions(clusterIP)
+	return ss.subscribe(port, listener)
+}
+
+// Unsubscribe removes a listener from the subscribers list for this authority.
+func (iw *IPWatcher) Unsubscribe(clusterIP string, port Port, listener EndpointUpdateListener) {
+	iw.log.Infof("Stopping watch on service cluster ip [%s:%d]", clusterIP, port)
+	ss, ok := iw.getServiceSubscriptions(clusterIP)
+	if !ok {
+		iw.log.Errorf("Cannot unsubscribe from unknown service ip [%s:%d]", clusterIP, port)
+		return
+	}
+	ss.unsubscribe(port, listener)
+}
+
+func (iw *IPWatcher) addService(obj interface{}) {
+	service := obj.(*corev1.Service)
+	if service.Namespace == kubeSystem || service.Spec.ClusterIP == "None" {
+		return
+	}
+
+	ss := iw.getOrNewServiceSubscriptions(service.Spec.ClusterIP)
+
+	ss.updateService(service)
+}
+
+func (iw *IPWatcher) deleteService(obj interface{}) {
+	service := obj.(*corev1.Service)
+	if service.Namespace == kubeSystem {
+		return
+	}
+
+	ss, ok := iw.getServiceSubscriptions(service.Spec.ClusterIP)
+	if ok {
+		ss.deleteService()
+	}
+}
+
+func (iw *IPWatcher) addPod(obj interface{}) {
+	pod := obj.(*corev1.Pod)
+	if pod.Namespace == kubeSystem {
+		return
+	}
+	if pod.Status.PodIP == "" {
+		// Pod has not yet been assigned an IP address.
+		return
+	}
+	ss := iw.getOrNewServiceSubscriptions(pod.Status.PodIP)
+
+	ownerKind, ownerName := iw.k8sAPI.GetOwnerKindAndName(pod, true)
```

Comment:
Won't `ss.pod` return the same thing? This seems like a repetition of what `iw.getOrNewServiceSubscriptions()` did to initialize the `PodSet` in the `serviceSubscription`, with the exception that `iw.getOrNewServiceSubscriptions()` also checks for `pod.spec.HostNetwork`.

Reply:
good call

Reply:
I think this change actually caused some problems because `getOrNewServiceSubscriptions` does not perform a pod lookup if the serviceSubscription already exists.  I've backed out this change, but refactored the common code into a `podToPodSet` method.